app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+import kotlin.coroutines.suspendCoroutine
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import org.osmdroid.util.BoundingBox
+
+/**
+ * View model to work with hikes.
+ *
+ * This view model is responsible for a central part of the app's functionality, it handles hikes.
+ *
+ * @param savedHikesRepo The repository to work with saved hikes.
+ * @param osmHikesRepo The repository to work with hikes from OpenStreetMap.
+ * @param elevationRepo The service to retrieve elevation data.
+ * @param dispatcher The dispatcher to be used to launch coroutines.
+ */
+class HikesViewModel(
+    private val savedHikesRepo: SavedHikesRepository,
+    private val osmHikesRepo: HikeRoutesRepository,
+    private val elevationRepo: ElevationService,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+) : ViewModel() {
+  companion object {
+    private const val LOG_TAG = "HikesViewModel"
+  }
+
+  private val _hikesMutex = Mutex()
+
+  private val _savedHikesMap = mutableMapOf<String, SavedHike>()
+
+  private var _hikeFlowsMap = mutableMapOf<String, MutableStateFlow<Hike>>()
+
+  private val _loading = MutableStateFlow(false)
+
+  private val _hikeFlowsList = MutableStateFlow<List<StateFlow<Hike>>>(emptyList())
+
+  private fun updateHikeFlowsList() {
+    _hikeFlowsList.value = _hikeFlowsMap.values.toList()
+  }
+
+  private var _selectedHikeId: String? = null
+  private val _selectedHike = MutableStateFlow<Hike?>(null)
+
+  /**
+   * Helper function to update the selected hike once (or before) [hikeFlows] has been updated.
+   *
+   * This function does not acquire the [_hikesMutex]. It is the responsibility of the caller to
+   * call this function inside of a [Mutex.withLock] block.
+   *
+   * This function does not switch context either, it is the responsibility of the caller to call
+   * this function inside of a [withContext] block.
+   */
+  private fun updateSelectedHike() {
+    // If there is no selected hike, don't bother.
+    if (_selectedHikeId == null) {
+      return
+    }
+
+    // Retrieve the selected hike from the map. If it's null, don't bother.
+    val selectedHike = _selectedHike.value ?: return
+
+    // Retrieve the corresponding flow from the map.
+    val selectedHikeFlow = _hikeFlowsMap[_selectedHikeId]
+
+    if (selectedHikeFlow == null) {
+      // The selected hike is not in the map, unselect it.
+      _selectedHikeId = null
+      _selectedHike.value = null
+    } else {
+      // The selected hike is still in the map, update it.
+      val flowValue = selectedHikeFlow.value
+      if (flowValue != selectedHike) {
+        _selectedHike.value = flowValue
+      }
```

Comment:
I can't answer for sure, but I'll tell you my reasons.

From what I read, when you call `_flow.value = newValue`, the flow will call all its subscribers for them to update based on the new state. This happens even when the previous state was the same.

Hence I did not want to cause unnecessary UI updates and potential performance issues, so I preferred to add this check here.

Now I never verified this theory in practice, because state flows are very hard to test (at least for what I know of them), so I can't tell you that this check changes anything for sure. But I think it's worth keeping nevertheless.

Reply:
That's what I was expecting, you confirmed what I thought, so let's keep it like this. I prefer to have one additional check than a UI update